### DISCUSSION TITLE
MATCH (v) RETURN v LIMIT n returns -1005:Storage Error: part: 40, error: E_RPC_FAILURE(-3).

### DISCUSSION THREAD

**Discussion by sunkararp** [vesoft-inc/nebula#4294] [Q&A]
Nebula version: 3.1.0

Hi @wey-gu simple below query fails with the error.

```
MATCH (v:Student) RETURN v LIMIT 10
```
Error
```
-1005:Storage Error: part: 40, error: E_RPC_FAILURE(-3)
```

There are about 10 million Student vertices in the system.

- Is my query timing out? 
- If timing out, per how to resolve [faq](https://docs.nebula-graph.io/3.1.0/20.appendix/0.FAQ/#how_to_resolve_the_error_storage_error_e_rpc_failure) I need to extend the connection timeout of the Storage client by setting --storage_client_timeout_ms=60000 in nebula-graphd.conf file
- Where is nebula-graphd.conf file and how I set --storage_client_timeout_ms in K8S environment?
- Running below command brings up all graphd severs
```
show hosts
```
- I tried to edit nebula-cluster-graphd, but my edits are overridden by resetting graphd
```
kubectl edit configmap nebula-cluster-graphd -n mynamespace
```

**Comment by wey-gu** [comment]
For configuration change in k8s operator, please refer to: https://github.com/vesoft-inc/nebula-operator/blob/master/doc/user/custom_config.md and in case it's not doable with this, we could raise it to nebula-operator repo as a request.

```
MATCH (v:Student) RETURN v LIMIT 10
```
In This query, limit is not pushdown to storage side, thus the query is fetching all data to graphd, which is quite costly actually.

You could for now try `lookup on Student yield vertex as v | limit 10` instead it'll be extremely fast :)

**Reply by sunkararp** [reply]
@wey-gu thanks for the reply

Lookup is very fast, thanks. Below is my actual query I was trying.
How I rewrite my query using look up?

```
MATCH (s:Student) 
WITH s LIMIT 30
MATCH (s)-[pdi:STUDENT_HAS_CLASS_CODE]-(di:Classcode)
WITH p, di LIMIT 100 
MATCH (di)-[pdi1:CLASS_CODE_HAS_STUDENT]-(s1:Student)
WHERE s.Student.studentId <> s1.Student.studentId
RETURN s.Student.studentId as source, s1.Student.studentId as destination
LIMIT 100
```

**Reply by wey-gu** [reply]
This single query could be written into(pipeline is your friend here):

```sql
LOOKUP ON Student yield id(vertex) AS student | LIMIT 30 |
  GO FROM $-.student OVER STUDENT_HAS_CLASS_CODE YIELD src(edge) AS src, dst(edge) AS shared_class | LIMIT 100 |
  GO FROM $-.shared_class OVER CLASS_CODE_HAS_STUDENT YIELD $-.src AS source, dst(edge) AS destination | LIMIT 100
```

For existing query, there are some points to be commented from my side:
- It's good to use WITH to enable multiple steps LIMIT 👍🏻
- WHEN possible, provide edge direction will help on performance like --> rather than --
- STUDENT_HAS_CLASS_CODE and CLASS_CODE_HAS_STUDENT are just one relationship in a different direction, this somehow is a duplicated edge, but I agree it's good for human readability :)
- pdi and pdi1 could be omit in this query

**Reply by sunkararp** [reply]
- this query took about 153 seconds consuming about 16 GB; can we further refine?
- the modified query uses LOOKUP & GO nGQL constructs
- My original open cypher query doesn't work

**Reply by Milittle** [reply]
> For configuration change in k8s operator, please refer to: https://github.com/vesoft-inc/nebula-operator/blob/master/doc/user/custom_config.md and in case it's not doable with this, we could raise it to nebula-operator repo as a request.
> 
> ```
> MATCH (v:Student) RETURN v LIMIT 10
> ```
> 
> In This query, limit is not pushdown to storage side, thus the query is fetching all data to graphd, which is quite costly actually.
> 
> You could for now try `lookup on Student yield vertex as v | limit 10` instead it'll be extremely fast :)

Is this limit will be supported to pushdown to sotrage, any plan? thx

**Reply by wey-gu** [reply]
> * this query took about 153 seconds consuming about 16 GB; can we further refine?

Do you mean the `lookup` one or the `match` one, please? If it's the lookup one, could you help share the profile of the query?

**Reply by wey-gu** [reply]
> >   
> > In This query, limit is not pushdown to storage side, thus the query is fetching all data to graphd, which is quite costly actually.
> > You could for now try `lookup on Student yield vertex as v | limit 10` instead it'll be extremely fast :)
> 
> Is this limit will be supported to push down to storage, any plan? thx

@MuYiYong do we have plans to support match limit pushdown in nebula repo, please :D

**Reply by Milittle** [reply]
I know when the tag do not have index, the match limit sentence will be pushdown to storage. but when create one index for tag, the RBO rule can not matched, so I think maybe adjust some code, it will be supported?

**Reply by hesy007** [reply]
we want to use cypher instead of lookup
